### ISSUE TITLE
multiple compilation errors due to library breaking changes - libtorrent-rasterbar 1:1.2.4-1

### ISSUE THREAD

**Issue by Manu-sh** [arvidn/libtorrent#4374]
This piece of code (and other always realted to *flags*) dosn't compile anymore how i should use the library now to obtain the same behavior?

```cpp
using enum_flags = tlib::file_storage::flags_t;
const static std::unordered_map<int,int> lookup_tab{
	{ enum_flags::attribute_hidden,     FileInfo::ATTR_HIDDEN },
	{ enum_flags::attribute_executable, FileInfo::ATTR_EXECUTABLE },
	{ enum_flags::attribute_symlink,    FileInfo::ATTR_SYMLINK }
};
```
my library version actually is: `extra/libtorrent-rasterbar 1:1.2.4-1` (archlinux)

```
main.cpp:122:41: error: ‘flags_t’ in ‘class libtorrent::file_storage’ does not name a type
  122 |  using enum_flags = tlib::file_storage::flags_t;
```

others errors always related to flags definitions are:
```
main.cpp:246:34: error: ‘category_t’ in ‘class libtorrent::alert’ does not name a type
  246 |  using alert_type = tlib::alert::category_t;
```

```
main.cpp:283:52: error: ‘flag_share_mode’ is not a member of ‘libtorrent::add_torrent_params::flags_t’ {aka ‘libtorrent::flags::bitfield_flag<long unsigned int, libtorrent::torrent_flags_tag>’}
  283 |  minfo.flags |= tlib::add_torrent_params::flags_t::flag_share_mode | tlib::add_torrent_params::flags_t::flag_upload_mode | tlib::add_torrent_params::flags_t::flag_super_seeding;
```

**Comment by arvidn** [comment]
both `category_t` and `flags_t` were moved out of their respective classes. I will add back aliases for these types to ease transition for libtorrent-1.2.5. This was an oversight. thanks!

Unfortunately, it's hard to be backwards compatible with the syntax in your 3rd example. If you just refer to the flags without the name of the enum, it should work. Like this:

```
minfo.flags |= tlib::add_torrent_params::flag_share_mode | tlib::add_torrent_params::flag_upload_mode | tlib::add_torrent_params::flag_super_seeding;
```

Thanks for reporting this!

**Comment by arvidn** [comment]
please give this patch a try if you can! https://github.com/arvidn/libtorrent/pull/4375

**Comment by Manu-sh** [comment]
Thank you maybe this week i could try, i do these things in my spare time, really appreciate your answer!

**Comment by arvidn** [comment]
great, thanks! I landed that patch, so you could pick it up from the main stable branch, `RC_1_2`. Please let me know if you find any other examples like this. I went over the changes since 1.1.x and added aliases for a few other flag types too.

**Comment by Manu-sh** [comment]
> great, thanks! I landed that patch, so you could pick it up from the main stable branch, `RC_1_2`. Please let me know if you find any other examples like this. I went over the changes since 1.1.x and added aliases for a few other flag types too.

i have build the library with `mkdir build && cd build && cmake -DCMAKE_INSTALL_PREFIX:PATH=/usr .. && sudo cmake --build . --target install --config Release`

i still get these errors (*in this moment i don't really need to fix it*) however FYU (if you need i can send the source file in pvt)
```
main.cpp:124:17: error: ‘attribute_hidden’ is not a member of ‘enum_flags’ {aka ‘libtorrent::flags::bitfield_flag<unsigned char, libtorrent::file_flags_tag>’}
  124 |   { enum_flags::attribute_hidden,     FileInfo::ATTR_HIDDEN },
      |                 ^~~~~~~~~~~~~~~~
main.cpp:125:17: error: ‘attribute_executable’ is not a member of ‘enum_flags’ {aka ‘libtorrent::flags::bitfield_flag<unsigned char, libtorrent::file_flags_tag>’}
  125 |   { enum_flags::attribute_executable, FileInfo::ATTR_EXECUTABLE },
      |                 ^~~~~~~~~~~~~~~~~~~~
main.cpp:126:17: error: ‘attribute_symlink’ is not a member of ‘enum_flags’ {aka ‘libtorrent::flags::bitfield_flag<unsigned char, libtorrent::file_flags_tag>’}
  126 |   { enum_flags::attribute_symlink,    FileInfo::ATTR_SYMLINK }
      |                 ^~~~~~~~~~~~~~~~~
main.cpp:127:2: error: no matching function for call to ‘std::unordered_map<int, int>::unordered_map(<brace-enclosed initializer list>)’
  127 |  };
      |  ^
```


```
main.cpp:283:52: error: ‘flag_share_mode’ is not a member of ‘libtorrent::add_torrent_params::flags_t’ {aka ‘libtorrent::flags::bitfield_flag<long unsigned int, libtorrent::torrent_flags_tag>’}
  283 |  minfo.flags |= tlib::add_torrent_params::flags_t::flag_share_mode | tlib::add_torrent_params::flags_t::flag_upload_mode | tlib::add_torrent_params::flags_t::flag_super_seeding;
      |                                                    ^~~~~~~~~~~~~~~
main.cpp:283:105: error: ‘flag_upload_mode’ is not a member of ‘libtorrent::add_torrent_params::flags_t’ {aka ‘libtorrent::flags::bitfield_flag<long unsigned int, libtorrent::torrent_flags_tag>’}
  283 |  minfo.flags |= tlib::add_torrent_params::flags_t::flag_share_mode | tlib::add_torrent_params::flags_t::flag_upload_mode | tlib::add_torrent_params::flags_t::flag_super_seeding;
      |                                                                                                         ^~~~~~~~~~~~~~~~
main.cpp:283:159: error: ‘flag_super_seeding’ is not a member of ‘libtorrent::add_torrent_params::flags_t’ {aka ‘libtorrent::flags::bitfield_flag<long unsigned int, libtorrent::torrent_flags_tag>’}
  283 |  minfo.flags |= tlib::add_torrent_params::flags_t::flag_share_mode | tlib::add_torrent_params::flags_t::flag_upload_mode | tlib::add_torrent_params::flags_t::flag_super_seeding;
      |                                                                                                                                                               ^~~~~~~~~~~~~~~~~~
main.cpp:284:52: error: ‘flag_upload_mode’ is not a member of ‘libtorrent::add_torrent_params::flags_t’ {aka ‘libtorrent::flags::bitfield_flag<long unsigned int, libtorrent::torrent_flags_tag>’}
  284 |  minfo.flags |= tlib::add_torrent_params::flags_t::flag_upload_mode;
      |                                                    ^~~~~~~~~~~~~~~~
main.cpp:298:71: error: ‘all_categories’ is not a member of ‘alert_type’ {aka ‘libtorrent::flags::bitfield_flag<unsigned int, libtorrent::alert_category_tag>’}
  298 |  sett.set_int(tlib::settings_pack::int_types::alert_mask, alert_type::all_categories); // TODO
      |                                                                       ^~~~~~~~~~~~~~
main.cpp:321:63: error: no matching function for call to ‘TorrentInfo::TorrentInfo(<brace-enclosed initializer list>)’
  321 |     TorrentInfo tinfo{thandle.torrent_file(), thandle.status()};
```

**Comment by arvidn** [comment]
those enums were not `enum class`, so they behave in somewhat strange ways. This is one of the reason I transitioned away from using enums. I think the most reasonable fix would be to change your code to not specify `flags_t` when naming flags. Just remove it, it will work in both the new and old API (since the enum name is optional for classing enums).

I'm not sure what `enum_flags` is in your first example. Is that a typedef you made of the type? Basically, the assumption that the *type* of the flags also contain the flags is not a safe assumption (and technically it was only implied by the documentation).

In short, don't name enums. Just use the names of the flags from the outer scope.

The last error in your first example is a consequence of the improved type safety. Flags are no longer `int`s.
